File: Frontend/DiveApp/Repositories/AuthRepository.swift
import Foundation

final class AuthRepository {
    private let authService: AuthService
    private let tokenStorage: TokenStorage

    init(authService: AuthService, tokenStorage: TokenStorage) {
        self.authService = authService
        self.tokenStorage = tokenStorage
    }

    /// Returns whether this is the user's first login (client should route to Profile Setup).
    @discardableResult
    func login(provider: AuthProvider, providerToken: String) async throws -> Bool {
        let tokens = try await authService.login(provider: provider, token: providerToken)
        tokenStorage.save(accessToken: tokens.accessToken, refreshToken: tokens.refreshToken)
        return tokens.isNewUser
    }

    func refreshAccessToken() async throws -> String {
        guard let refreshToken = tokenStorage.refreshToken else {
            throw APIError.unauthorized
        }
        let response = try await authService.refresh(refreshToken: refreshToken)
        tokenStorage.accessToken = response.accessToken
        return response.accessToken
    }

    func logout() async {
        try? await authService.logout()
        tokenStorage.clear()
    }

    var hasStoredSession: Bool {
        tokenStorage.refreshToken != nil
    }

    var currentAccessToken: String? {
        tokenStorage.accessToken
    }
}

File: Android/app/src/main/java/com/diveapp/android/repository/AuthRepository.kt
package com.diveapp.android.repository

import com.diveapp.android.core.auth.TokenStorage
import com.diveapp.android.core.network.ApiException
import com.diveapp.android.core.network.apiCall
import com.diveapp.android.model.AuthProvider
import com.diveapp.android.model.RefreshTokenRequest
import com.diveapp.android.model.SocialLoginRequest
import com.diveapp.android.service.AuthService

class AuthRepository(
    private val authService: AuthService,
    private val tokenStorage: TokenStorage,
) {
    /** Returns whether this is the user's first login (caller should route to Profile Setup). */
    suspend fun login(provider: AuthProvider, providerToken: String): Boolean {
        val tokens = apiCall { authService.login(provider.name, SocialLoginRequest(providerToken)) }
        tokenStorage.save(tokens.accessToken, tokens.refreshToken)
        return tokens.isNewUser
    }

    suspend fun refreshAccessToken(): String {
        val refreshToken = tokenStorage.refreshToken ?: throw ApiException.Unauthorized
        val response = apiCall { authService.refresh(RefreshTokenRequest(refreshToken)) }
        tokenStorage.accessToken = response.accessToken
        return response.accessToken
    }

    suspend fun logout() {
        runCatching { apiCall { authService.logout() } }
        tokenStorage.clear()
    }

    /** Clears local session state without calling the API — used when the
     * server has already rejected the refresh token (see TokenAuthenticator). */
    fun clearLocalSession() {
        tokenStorage.clear()
    }

    val hasStoredSession: Boolean
        get() = tokenStorage.refreshToken != null
}
